### PR TITLE
Machines: fix DNS and DHCP related issues

### PR DESCRIPTION
**Pull request by vasile-papp**
While using Cockpit 211.3 with CentOS 8 I noticed issues which I was still able to identify in the latest version in the master branch.
Please see the long description of each commit for the reasoning behind them.
Thank you for your consideration,
Vasile

### CODE REVIEW

**Comment by KKoukiou** [comment]
@vasile-papp thanks for the contribution, we will review in the following days.

File: pkg/machines/xmlCreator.js
Comment (skobyda) [inline pkg/machines/xmlCreator.js:88]: From libvirt documentation:
`If the optional localOnly attribute on the domain element is "yes", then DNS requests under this domain will only be resolved by the virtual network's own DNS server - they will not be forwarded to the host's upstream DNS server. If localOnly is "no", and by default, unresolved requests will be forwarded.`

Wouldn't this resolve only in limiting us? Since we would be only be limited on virtual network's DNS for domain translation, and if no entry is found there, then request is not forwarded? Correct me if I'm wrong, but wouldn't this result in lost of domain translation for Internet?

File: pkg/machines/xmlCreator.js
Comment (skobyda) [inline pkg/machines/xmlCreator.js:97]: Could you please fix this typo in a commit where you introduced it? Since it's in same Pull request, there is no need for typos made in one commit to be fixed another commit. (rebase and fixup should do the job)

File: pkg/machines/xmlCreator.js
Comment (vasile-papp) [reply pkg/machines/xmlCreator.js:88]: The network name is used by the virtual DNS, for example "default", "local", "intranet" and most probably not like "google.com". By default we have the "default" network & domain, so any "vm" would be registered as "vm.default" in the virtual DNS. Request for "vm.default" and "inexistent.vm" would be dealt with by the virtual DNS  without external traffic, and "www.google.com" would be also resolved by the virtual DNS by forwarding upstream the query.
In my setup, I have my own DNS server configured to forward the "default" domain requests to the virtual DNS, so it can resolve on my network both internet and vm names. In case the "localOnly" option is not used, there is a loop between the DNS server and the virtual DNS for trying to resolve the "inexistentvm.default" or the "stoppedvm.default", because the virtual DNS will forward all not known requests to the upstream DNS which in this case if my main DNS server that will send the request back to the virtual DNS since it's related to "default" domain.
Once the "localOnly" option is used, then the virtual DNS will know that is the only DNS capable to solve hosts from "default" domain (existent or not), any other domains would be forwarded upstream for resolution.
I mentioned in the commit details the other scenario if we want to have network names like "google.com", which looks to be a nice hack, but is it a desired feature to have? if yes, then might be an idea to have a checkbox for using "localOnly".
(any time it's possible to add it by editing the network outside the cockpit interface, not handy for everybody)

File: pkg/machines/xmlCreator.js
Comment (vasile-papp) [reply pkg/machines/xmlCreator.js:97]: Done. Thanks for the hint, I didn't know before how to do that.

File: pkg/machines/xmlCreator.js
Comment (skobyda) [reply pkg/machines/xmlCreator.js:88]: > In my setup, I have my own DNS server configured to forward the "default" domain requests to the virtual DNS

But this needs to be configured extra, right? And would this still work even if the user doesn't have their DNS server set up that way? It's not good to introduce "localOnly=yes" option as a default setting if it needs some additional configuration to make it work.
Maybe allowing user to optionally change virtual network as localOnly should be better solution.

Comment (skobyda) [reply pkg/machines/xmlCreator.js:88]: > I mentioned in the commit details the other scenario if we want to have network names like "google.com", which looks to be a nice hack, but is it a desired feature to have?

I don't think it's big use case. I have not seen any user wishing to use virtual networks on Cockpit (or libvirt) for this.

File: pkg/machines/xmlCreator.js
Comment (skobyda) [inline pkg/machines/xmlCreator.js:95]: Here you have a typo. Instead of setAttribute you use setAtrribute.
```
setAttribute
setAtrribute
```

File: pkg/machines/xmlCreator.js
Comment (skobyda) [inline pkg/machines/xmlCreator.js:98]: Also a typo. Instead of appendChild you use appenChild.
```
appendChild
appenChild
```

**Review by skobyda — CHANGES_REQUESTED** [review]

File: pkg/machines/xmlCreator.js
Comment (vasile-papp) [reply pkg/machines/xmlCreator.js:88]: No extra configuration is needed, I only did that because I wanted it to. Usual requirement is to have the VM to resolve internet names and other VM names from same virtual network, both of which are not affected negatively by the 'localOnly=yes'.
In fact it improves resolution of non-existing or stopped VM names.

Comment (zippy2) [reply pkg/machines/xmlCreator.js:88]: > No extra configuration is needed, I only did that because I wanted it to. Usual requirement is to have the VM to resolve internet names and other VM names from same virtual network, both of which are not affected negatively by the 'localOnly=yes'.

I think you need to either configure network so that in DHCP it tells VMs to use DNS of the resolving VM, or configure VMs statically. And the resolving VM has to have two interfaces, right? One connected to the network for which it's doing the DNS translation and the other connected to the outside world (e.g. a macvtap or another network).

> In fact it improves resolution of non-existing or stopped VM names.

Yeah, I can imagine that this offers more options to play with than what libvirt provides. But stopped VMs will also be translated by libvirt network as long as the DHCP lease doesn't time out.

localOnly is what you want, sure. But only for this narrow use case. You definitely do not want to turn it on by default for all networks and IIUC that is what Simon is objecting to.

Comment (vasile-papp) [reply pkg/machines/xmlCreator.js:88]: > I think you need to either configure network so that in DHCP it tells VMs to use DNS of the resolving VM, or configure VMs statically. And the resolving VM has to have two interfaces, right? One connected to the network for which it's doing the DNS translation and the other connected to the outside world (e.g. a macvtap or another network).

No, I think there is a confusion between "local only network" (isolated) and "local only domain". There is no need to have 2 interfaces, because "localOnly" is not influencing the traffic rules, the network type (ex. NAT) would still function as expected.
The "localOnly" option refers only to the way how DNS will handle requests for the specified domain name, for example in case of "default" network the domain name is "default" which would would mean that the following line:
`local=/default/`
would be added in the `/var/lib/libvirt/dnsmasq/default.conf` file,
Based on `dnsmasq` config documentation means:
`    # Add local-only domains here, queries in these domains are answered`
`    # from /etc/hosts or DHCP only.`
`    #local=/localnet/`

Exactly, name resolution is OK until DHCP leash expires, then name resolution would encounter the mentioned problem.

My scenario might look "narrow use case" it is just a scenario in which the impact is more visible so a root cause analysis was the way to go, but the problem of name resolution is not specific.

The simplest use case that can reproduce the problem is to have 2 VMs `vm1` and `vm2` that should resolve each others name, after a host restart boot just one of them `vm1` (just imagine that vm2 failed to start or that it crashed and lease expired) and try to resolve `vm2` name while monitoring the DNS queries on the host with `tcpdump -i any -n port 53 | grep vm2`. 
In the output there will be queries to the Virtual DNS and to the upstream the DNS configured on the host (one of: router DNS, ISP DNS, Google DNS etc.)
Using the "localOnly=yes" would eliminate from that chain the upstream DNS configured on the host, ie. `vm2` and `vm2.default` would results in immediate reply from the virtual DNS regardless of the state of `vm2`.
Everything else would work in the same way.

My questions are:
 - is it OK to have VM names "leak" to ISP/Internet ? unnecessary traffic, small but why not eliminate it if not needed.
 - is it common to resolve "somename" in a VM and expect it to be handled by upstream DNS to "somename.site.com" or "somename" on the physical network? to reach internet full names are generally used, short names are expected to be names on local network (virtual network in this case).

**Comment by skobyda** [comment]
Sorry for getting to this so late. So far it seems this can get merged (libvirt are okay with that too).
It would be nice to have some tests for this fix. I can write the tests for this. @vasile-papp  Can you just give me examples of how to reproduce the action which this PR fixes, actions which didn't work before but after this would get merged it would work? (writing steps of recreating it in the UI, or command line, or combination of both is sufficient) With that, I can come up with some tests.

**Comment by vasile-papp** [comment]
The issues are valid when using Virtual Network with NAT forward mode.


1. IP should not be same as network or broadcast.

This commit is actually a test for the values entered in the Web GUI when creating the Virtual Network.
A validation method would be to provide different IP/Mask combinations to the function:
    - 1.1.1.0/24 should resturn False (IP/Mask is network addres)
    - 1.1.1.1/24 should return True (IP/Mask is host address)
    - 1.1.1.255/24 should return False (IP/Mask is broadcast address)
or using a more "complex" netmask:
    - 1.1.1.64/26 should return Flase (network)
    - 1.1.1.65/26 should return True (host)
    - 1.1.1.127/26 should return Flase (broadcast)


2. 192.168.100.0/24 it's a reserved network address

This commit is just an initial value which is valid host address (based on the test above) usable as gateway on the virtual network.
The user might change it and then the test above is validating the users choice.


3. Treat all DNS A record requests for this domain

The purpose is for Virtual Network that has it's own virtual DNS, to resolve names without forwarding to other parent DNS.
A possible testing scenario would be to:
    - have a parent DNS configured to solve "testhost.default" to an IP
    - have Virtual DNS for "default" network with "testvm.default" resolvable
    - query the Virtual DNS with "testhost.default" name, should return unknown (as there is no "testhost" VM in the virtual network, proving that the query was not forwarded)
    - query the Virtual DNS with "testvm.default" name, should return an IP (proving that the virtual DNS is able to resolve names of VMs from the virtual network)


4. provided IP resolvable by DNS

This commit only adds a name for the IP defined in the WebGUI as network gateway (192.168.100.1).
To verify that is OK, the Virtual DNS on "default" virtual network should be able to resolve the IP 192.168.100.1 to "gateway.default" name and the name back to the IP. Otherwise it's not working.


5. DNS PTR record for this virtual IP network

This commit would resolve IPs back to names without querying the parent DNS.
A possible testing scenario would be to:
    - have a parent DNS configured to solve "192.168.100.100" to an "testhost" name
    - have Virtual DNS for "default" network with "192.168.100.199" resolvable to "testvm"
    - query the Virtual DNS with "192.168.100.100" IP, should return unknown (as there is no "192.168.100.100" VM in the virtual network, proving that the query was not forwarded)
    - query the Virtual DNS with "192.168.100.199" IP, should return a name "testvm.default" (proving that the virtual DNS is able to resolve IPs of VMs from the virtual network)

**Comment by skobyda** [comment]
I wrote few tests for that. Could you give me push right to your this branch @vasile-papp so I could push my commits to this PR?

**Comment by vasile-papp** [comment]
> I wrote few tests for that. Could you give me push right to your this branch @vasile-papp so I could push my commits to this PR?

Invited you as collaborator.

**Comment by skobyda** [comment]
Added few tests. Also I reworder your commit messages (added "machines:" prefix, fixed line breaking and added "Closes: #14564" at last commit). If tests pass, I'm okay with merging it

**Comment by marusak** [comment]
I actually triggered those tests (@skobyda FYI you need to `tests-trigger --allow`) and it fails a lot:
```
TypeError: 'NoneType' object is not iterable
```

File: pkg/machines/xmlCreator.js
Comment (KKoukiou) [inline pkg/machines/xmlCreator.js:96]: Can someone explain this here? 
I usually use this xml syntax to assign static IPs to VMs with specific hostname, so where you put the gateway I usually put the VM's hostname? 
I am sure this here is something different, @skobyda or @vasile-papp if you can please clarify this, otherwise it's good to get merged.

File: pkg/machines/xmlCreator.js
Comment (KKoukiou) [inline pkg/machines/xmlCreator.js:97]: Can someone explain this here?
I usually use this xml syntax to assign static IPs to VMs with specific hostname, so where you put the gateway I usually put the VM's hostname?
I am sure this here is something different, @skobyda or @vasile-papp if you can please clarify this, otherwise it's good to get merged.

Comment (vasile-papp) [reply pkg/machines/xmlCreator.js:97]: Yes, the "dns" section is exactly for that purpose (associate "name" to "IP" or "IP" to "name"), this piece of code will just assign the "gateway" name to the "IP" address of the default gateway from the virtual network (in order to be resolved by the virtual dns).
```
  <dns>
    <host ip='192.168.100.1'>
      <hostname>gateway</hostname>
    </host>
  </dns>
```
You can continue to add new "hosts" in this section just as before.
```
  <dns>
    <host ip='192.168.100.1'>
      <hostname>gateway</hostname>
    </host>
    <host ip='192.168.100.100'>
      <hostname>vm1</hostname>
    </host>
    <host ip='192.168.100.123'>
      <hostname>vm2</hostname>
    </host>
  </dns>
```

Comment (KKoukiou) [reply pkg/machines/xmlCreator.js:97]: So basically in your machine you have a VM called gateway, and you want that IP assigned statically. What I don't understand is why should that we the default configuration for everyone using cockpit-machines?

Would you mind joining #cockpit on freenode so that we can finalize the discussion and if all clear merge this today?

Thanks in advance for your contribution and sorry it took ages to get this in.

**Review by KKoukiou — APPROVED** [review]

**Pull request merged**
